llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -7929,16 +7930,30 @@ LegalizerHelper::lowerThreewayCompare(MachineInstr &MI) {
                                        ? CmpInst::Predicate::ICMP_SGT
                                        : CmpInst::Predicate::ICMP_UGT;
 
-  auto One = MIRBuilder.buildConstant(DstTy, 1);
   auto Zero = MIRBuilder.buildConstant(DstTy, 0);
   auto IsGT = MIRBuilder.buildICmp(GTPredicate, CmpTy, Cmp->getLHSReg(),
                                    Cmp->getRHSReg());
-  auto SelectZeroOrOne = MIRBuilder.buildSelect(DstTy, IsGT, One, Zero);
-
-  auto MinusOne = MIRBuilder.buildConstant(DstTy, -1);
   auto IsLT = MIRBuilder.buildICmp(LTPredicate, CmpTy, Cmp->getLHSReg(),
                                    Cmp->getRHSReg());
-  MIRBuilder.buildSelect(Dst, IsLT, MinusOne, SelectZeroOrOne);
+
+  auto &Ctx = MIRBuilder.getMF().getFunction().getContext();
+  const DataLayout &DL = MIRBuilder.getDataLayout();
+  if (TLI.shouldExpandCmpUsingSelects(
```

Comment:
The condition for the select version slightly differs in the original PR:
`shouldExpandCmpUsingSelects() || BoolVT.getScalarSizeInBits() == 1 ||
      getBooleanContents(BoolVT) == UndefinedBooleanContent)`

Reply:
Doesn't the use of i1 for the cmp type avoid needing the other checks?

Reply:
Never mind it doesn't

Reply:
I added the equivalent of getBooleanContents(BoolVT) == UndefinedBooleanContent but not BoolVT.getScalarSizeInBits() == 1. I can't do it without having a setcc result type info.

Reply:
I'm actually worried the lack of setcc equivalent here. For SVE for example we return i32 for scalars (fine) but <n x M x i1> for scalable predicates, which would be incorrect if used with these intrinsics.

Maybe we should a wrapper around `getSetCCResultType()` that supports LLTs and just calls the MVT version? @arsenm thoughts?

Reply:
> I'm actually worried the lack of setcc equivalent here. For SVE for example we return i32 for scalars (fine) but `<vscale x N x i1>` for scalable predicates, which would be incorrect if used with these intrinsics.
> 
> 
> 
> Maybe we should a wrapper around `getSetCCResultType()` that supports LLTs and just calls the MVT version? @arsenm thoughts? 

I'm not sure I follow what would be incorrect. The G_ICMP we create here will use i1 results and go through legalization. Nothing should be wrong just not optimal.

Reply:
> > I'm actually worried the lack of setcc equivalent here. For SVE for example we return i32 for scalars (fine) but `<vscale x N x i1>` for scalable predicates, which would be incorrect if used with these intrinsics.
> > Maybe we should a wrapper around `getSetCCResultType()` that supports LLTs and just calls the MVT version? @arsenm thoughts?
> 
> I'm not sure I follow what would be incorrect. The G_ICMP we create here will use i1 results and go through legalization. Nothing should be wrong just not optimal.

Looking at the source SDAG comment:
> // We can't perform arithmetic on i1 values.

I guess the issue is that if your setcc result type is -1, you end up needing to be able to generate `0b11` but your `SUB` inputs will still be 1-bit booleans?

Reply:
This patch extends the bools to the scmp/ucmp result type, which is at least i2, before the subtract. So there shouldn't be a correctness issue there.

Reply:
Ah, I see. So for SVE this will cause some bad codegen but I guess we can deal with that when we need to. Can you at least put a comment in the code explaining why you're doing the extension? I think in future we may want to have a similar hook to `getSetCCResultType()`.

Reply:
Yes, we will need getSetCCResultType eventually. For AMDGPU this will need to be register bank aware after regbankselect